OneCommunityGlobal/HighestGoodNetworkApp
# Aishwarya - Purchase tool request form-frontend

Description
Adds Purchase Request: Tool page and form to the BM Dashboard (Phase 2).

Related PRS (if any):
Backend PR 670 must be running to test this branch.

Main changes explained:
Adds an ability to initiate a purchase request for a tool to be used on a specific project.
request must include quantity and priority level, with an optional 'preferred brand' field.
adds actions and reducers to store tool types data in redux store
adds error handling to fetch request on page load
adds success and error toasts after submitting purchase request

How to test:
follow checkout instructions for backend PR 670
git checkout Aish purchase tool request form (frontend)
npm install
npm run start:local
log in to HGN as a volunteer user
navigate to http://localhost:3000/bmdashboard and enter credentials again
navigate to http://localhost:3000/bmdashboard/tools/purchase

Test basic form submission:
Project and Tools options should auto-load into respective fields
Enter info into all required fields (Preferred Brand is optional) and click Submit
Note whether a success toast notification is received
Optional: check Network tab in browser tools to check for a 'tools' request with status code 201 (Created).

Test form validation:
Test that Submit button is disabled unless all required fields are entered
Test that entering a decimal or technically-allowable letter 'e' (ex: 10e10) in Quantity field results in a validation check/error.
Note: there are validation checks for all form fields, but these are harder to test without manipulating the HTML.
Test that Cancel button clears form and returns user to previously visited page.

Test error handling:
- To test error handling on initial page load, you can deactivate the backend server and reload the page. This will load an error page with code 503.
- To test error handling if form submission fails, change the fetch route in src/actions/bmdashboard/toolsActions > purchaseTools from ENDPOINTS.BM_TOOLS to ENDPOINTS.BM_TOOLS + "s". This should result in an error toast with code 404.

Screenshots or videos of changes:
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91091456/b29f0efa-f9a0-49ff-86c8-4fba967bf969)


Notes:
Please remember to either Approve or Request Changes. Do not just leave a comment.
Please leave a review for BOTH front and backend branches after you have finished testing!
Permissions are currently not being built into the Phase 2 codebase. All functionality should be accessible to volunteer users.
The form will eventually only be accessible to Building Manager users. Purchase requests will be limited to the BM's own projects. For now these limitations are not in place.
